nsis/src/nsis/file/NsisExecutable.java

```diff
@@ -35,49 +42,117 @@ public NsisExecutable() {
-				.create(NsisExecutable.class);
+	public static NsisExecutable createNsisExecutable(GenericFactory factory, ByteProvider bp,
+			SectionLayout layout) throws IOException, InvalidFormatException {
+		NsisExecutable nsisExecutable = (NsisExecutable) factory.create(NsisExecutable.class);
 		nsisExecutable.initNsisExecutable(factory, bp, layout);
 		return nsisExecutable;
 	}
 
-	private void initNsisExecutable(GenericFactory factory, ByteProvider bp,
-			SectionLayout layout) throws IOException, InvalidFormatException {
-		this.reader = new FactoryBundledWithBinaryReader(factory, bp,
-				/* isLittleEndian= */ true);
+	private void initNsisExecutable(GenericFactory factory, ByteProvider bp, SectionLayout layout)
+			throws IOException, InvalidFormatException {
+		this.reader = new FactoryBundledWithBinaryReader(factory, bp, /* isLittleEndian= */ true);
 		this.headerOffset = findHeaderOffset();
-		initScriptHeader(bp);
+		initScriptHeader();
+		if ((this.scriptHeader.compressedHeaderSize & 0x80000000) != 0) { // Check
+																			// if
+																			// MSB
+																			// is
+																			// set
+			decompressData();
+		}
 	}
 
 	private long findHeaderOffset() throws IOException, InvalidFormatException {
-		for (long headerOffset = 0; headerOffset
-				+ NsisConstants.NSIS_MAGIC.length <= reader
-						.length(); headerOffset++) {
+		for (long headerOffset = 0; headerOffset + NsisConstants.NSIS_SIGINFO.length
+				+ NsisConstants.NSIS_MAGIC.length <= reader.length(); headerOffset++) {
 			byte[] content = reader.readByteArray(headerOffset,
-					NsisConstants.NSIS_MAGIC.length);
-			if (Arrays.equals(NsisConstants.NSIS_MAGIC, content)) {
-				return headerOffset;
+					NsisConstants.NSIS_SIGINFO.length + NsisConstants.NSIS_MAGIC.length);
+			if (Arrays.equals(Bytes.concat(NsisConstants.NSIS_SIGINFO, NsisConstants.NSIS_MAGIC),
+					content)) {
+				return headerOffset - StructConverter.DWORD.getLength(); // Include
+																			// flags
+																			// in
+																			// header
 			}
 		}
 		throw new InvalidFormatException("Nsis magic not found.");
 	}
 
-	private void initScriptHeader(ByteProvider bp)
-			throws IOException, InvalidFormatException {
-		BinaryReader br = new BinaryReader(bp, /* isLittleEndian= */ true);
-		br.setPointerIndex(this.headerOffset);
-		this.scriptHeader = new NsisScriptHeader(br);
+	private void initScriptHeader() throws IOException, InvalidFormatException {
+		this.reader.setPointerIndex(this.headerOffset);
+		this.scriptHeader = new NsisScriptHeader(this.reader);
+	}
+
+	/**
+	 * Attempt to decompress the data from the reader. Supports LZMA algorithm. Will
+	 * eventually support Bzip2 and Zlib.
+	 * 
+	 * @throws IOException
+	 */
+	private void decompressData() throws IOException {
+		byte compressionByte = this.reader.readNextByte();
+		long compressedDataOffset;
+		int compressedDataLength;
+		byte[] compressedData;
```

Comment:
Please declare variables where they are first used, not in the beginning of the function.
https://google.github.io/styleguide/javaguide.html#s4.8.2-variable-declarations

Reply:
Even if I will need those variables in the different if-else ?